zlib: update for latest zig api and add build.zig.zon

rockorager opened the pull request:
Update all zlib related calls to conform to the new API. Add a build.zig.zon which is required by the package manager now.

mlarouche commented:
@rockorager Hi, can you made that the test suite passes? Let me know if you need any help setting up the environment, see the instructions in the [README](https://github.com/zigimg/zigimg?tab=readme-ov-file#test-suite). Thanks :)

rockorager commented:
Ah I completely missed that there was a separate test suite! Sorry about that.

rockorager commented:
I'm a little stuck...I fixed the error reporting failures but there is an endianess test failure and I can't seem to figure out what is causing this...do you have any ideas?

I modified the failing call in `reader.zig` at 222 to log the error, and it's an `error.EndOfStream`...grepping through my changes (after the fix), I haven't *added* any EndOfStream errors, hence my confusion.

```
/home/tim/repos/zigimg/src/utils.zig:140:21: 0x11bc7dd in readStructForeign__anon_8879 (zigimgtest)
    var result: T = try reader.readStruct(T);
                    ^
/home/tim/repos/zigimg/src/formats/png/reader.zig:222:13: 0x1190849 in loadWithHeader (zigimgtest)
            return err;
            ^
/home/tim/repos/zigimg/src/formats/png/reader.zig:179:21: 0x1193ab0 in load (zigimgtest)
    result.pixels = try loadWithHeader(stream, &header, allocator, options);
                    ^
/home/tim/repos/zigimg/tests/formats/png_test.zig:218:26: 0x1192a62 in test.InfoProcessor on Png Test suite (zigimgtest)
            var result = try png.load(&stream, std.testing.allocator, options.get());
                         ^
error: while executing test 'test.filtering 16-bit grayscale pixels uses correct endianess', the following test command failed:
/home/tim/repos/zigimg/zig-out/bin/zigimgtest --listen=-
Build Summary: 3/5 steps succeeded; 1 failed; 166/169 tests passed; 2 skipped; 1 failed (disable with --summary none)
test transitive failure
└─ run zigimgtest 166/169 passed, 1 failed, 2 skipped
error: the following build command failed with exit code 1:
/home/tim/repos/zigimg/zig-cache/o/b9ae42cb7f47a6672fa3fccf55a1d123/build /home/tim/.zig/versions/zig-linux-x86_64-0.12.0-dev.2823+955fd65cb/zig /home/tim/repos/zigimg /home/tim/repos/zigimg/zig-cache /home/tim/.cache/zig --seed 0xb0a29e6b -Z53e6f43b519ae286 test
```

igor84 commented:
It seems there is a bug in the new decompressor. This is what happens in "InfoProcessor on Png Test suite":

1. The code is loading a png chunk at a time and it encounters a zTXt type of chunk with length 65. The chunk contents starts at stream offset 144 and ends on offset 209. Since every chunk has additional 4 bytes for the checksum after the contents that means that the next chunk is starting at offset 213.
2. The relevant code in InfoProcessor that handles zTXt chunk looks like this:
```rs
var decompressStream = std.compress.zlib.decompressor(reader);
var print_buf: [1024]u8 = undefined;
var got = decompressStream.read(print_buf[0..]) catch return error.InvalidData;
while (got > 0) {
	self.writer.print("{s}", .{print_buf[0..got]}) catch return result_format;
    got = decompressStream.read(print_buf[0..]) catch return error.InvalidData;
}
```
3. After the first call to `decompressStream.read` stream position is not 209 but 205 but the entire expected content is decompressed and read.
4. Then inside the while loop the second call to `decompressStream.read` returns that it read the expected 0 bytes but the position of the underlying stream is moved by 8 bytes and is now 213. So it is now four bytes after the end of the contents.
5. The code at the end of method calls additional `skipBy(4)` to skip the 4 checksum bytes but since the previous call already skipped them we are now in invalid position in the stream.
6. When the next chunk header is loaded we now get invalid huge number as its length which makes the code jump to the end of the file and the next read operation then fails with `EndOfStream` error.

This should be fixed in the decompressor code but in the meantime we can work around it by adding this at the beggining of the `processChunk` method in InfoProcessor.zig: `const endPos = try data.stream.getPos() + data.chunk_length + 4;` and then replacing the line bellow big switch statement that says `try data.stream.seekBy(@sizeOf(u32));` with `try data.stream.seekTo(endpos);`. At that point we can also remove all `data.stream.seekBy` calls in entire switch statement except the one in `png.Chunks.zTXt.id` branch.

rockorager commented:
Nice find! Do you know if zig is aware of this bug? I know they are closing in
on a 0.12.0 release - would be good to have this fixed for that.

andrewrk commented:
Perhaps this? https://github.com/ziglang/zig/issues/18967

igor84 commented:
@andrewrk it does sound like this issue. The only additional thing I noticed here that isn't mentioned is that the first read call, although returning the entire decompressed content actually didn't read the last 4 bytes from the compressed stream. It could be that it is also just some checksum or something like that so is not necessary for actually returning decompressed data but I would expect the first read call to leave the underlying stream at the position after those 4 bytes and the second call not to move it at all.

andrewrk commented:
I suggest to have a chat with @ianic on that issue even though it's closed - he's been pretty active and helpful lately.

ianic commented:
Of course, happy to help I probably make the problem in the first place. 

I checked out https://github.com/rockorager/zigimg/tree/zlib but tests are passing although there are lots of them skipped. 
Is there a way to get failing test so I can try understand what is exactly happening. 

Difference with previous implementation is that this has 8 bytes internal buffer. Resulting that at the end of stream some data can be in that buffer. For zlib we can expect 4 bytes or less of overshoot (after compressed data it reads 4 bytes of zlib checksum). There is method `unreadBytes` at decompressor to get exact number of overshoot bytes.

ianic commented:
Found 3 failing tests.

One could be eliminated by returning unread bytes to the stream, [here](https://github.com/rockorager/zigimg/blob/19a49a7e44fb4b1c22341dfbd6566019de742055/src/formats/png/InfoProcessor.zig#L105)
```
try data.stream.seekBy(-@as(i64, @intCast(decompressStream.unreadBytes())));
```
The other one by adding `error.EndOfStream` to the error set, [here](https://github.com/rockorager/zigimg/blob/19a49a7e44fb4b1c22341dfbd6566019de742055/src/formats/png/reader.zig#L453)

I'm left with this one: 
```
expected 16711680, found 16777215
/usr/local/zig/zig-linux-aarch64-0.12.0-dev.3033+031f23117/lib/std/testing.zig:93:17: 0x11604c7 in expectEqualInner__anon_6348 (zigimgtest)
                return error.TestExpectedEqual;
                ^
/home/ianic/zig/zigimg/tests/helpers.zig:15:5: 0x1161487 in expectEq__anon_6361 (zigimgtest)
    try testing.expectEqual(@as(@TypeOf(actual), expected), actual);
    ^
/home/ianic/zig/zigimg/tests/formats/tga_test.zig:969:13: 0x11c71d3 in test.Write compressed 16-bit true color TGA (zigimgtest)
            try helpers.expectEq(pixels.rgb555[stride + x].toU32Rgb(), expected_strip[strip_index]);
            ^
error: while executing test 'src.formats.png.filtering.test.filtering 16-bit grayscale pixels uses correct endianess', the following test command failed:
/home/ianic/zig/zigimg/zig-out/bin/zigimgtest --listen=-
Build Summary: 3/5 steps succeeded; 1 failed; 166/169 tests passed; 2 skipped; 1 failed (disable with --summary none)
test transitive failure
mq run zigimgtest 166/169 passed, 1 failed, 2 skipped
error: the following build command failed with exit code 1:
/home/ianic/zig/zigimg/zig-cache/o/082c51d891332ecb2f25c6effdb1d384/build /usr/local/zig/zig-linux-aarch64-0.12.0-dev.3033+031f23117/zig /home/ianic/zig/zigimg /home/ianic/zig/zigimg/zig-cache /home/ianic/.cache/zig --seed 0xe89e70da -Z7fa3314982d5f085 test
```

igor84 commented:
The API design where users have to be aware of and use `unreadBytes` feels very wrong, like something that will constantly trip the users up. Anyway I  extracted the problematic data into a separate test bellow so you can use it directly in the compressor lib @ianic:

```rs
test "Compress stream position test" {
    const len: usize = 58;
    const compressed_with_four_extra_bytes = [len]u8{
        0x78, 0x9c, 0x73, 0xce, 0x2f, 0xa8, 0x2c, 0xca, 0x4c, 0xcf, 0x28, 0x51, 0x08, 0xcf, 0xcc, 0xc9,
        0x49, 0xcd, 0x55, 0x28, 0x4b, 0xcc, 0x53, 0x08, 0x4e, 0xce, 0x48, 0xcc, 0xcc, 0xd6, 0x51, 0x08,
        0xce, 0xcc, 0x4b, 0x4f, 0x2c, 0xc8, 0x2f, 0x4a, 0x55, 0x30, 0xb4, 0xb4, 0x34, 0xd5, 0xb5, 0x34,
        0x03, 0x00, 0x8b, 0x61, 0x0f, 0xa4, 0x52, 0x5a, 0x94, 0x12
    };

    var base_stream = std.io.fixedBufferStream(compressed_with_four_extra_bytes[0..]);

    try std.testing.expectEqual(0, base_stream.getPos());
    var decompress_stream = std.compress.zlib.decompressor(base_stream.reader());
    var print_buf: [1024]u8 = undefined;

    try std.testing.expectEqual(8, base_stream.getPos());
    var got = try decompress_stream.read(print_buf[0..]);
    try std.testing.expectEqual(46, got);
    try std.testing.expectEqualStrings("Copyright Willem van Schaik, Singapore 1995-96", print_buf[0..got]);
    try std.testing.expectEqual(len - 4, base_stream.getPos()); // What I really expect
    // try std.testing.expectEqual(len - 8, base_stream.getPos()); // What I actually get

    got = try decompress_stream.read(print_buf[0..]);
    try std.testing.expectEqual(0, got);
    try std.testing.expectEqual(len - 4, base_stream.getPos()); // What I really expect
    // try std.testing.expectEqual(len, base_stream.getPos()); // What I actually get
}
```

ianic commented:
> The API design where users have to be aware of and use unreadBytes feels very wrong, like something that will constantly trip the users up. 

Completely agree.
I didn't think about usages where compressed stream in followed with some other data. When decompressing single stream of compressed data, like tar.gz archives overshoot is not visible.
It's not hard to change behavior for zlib, just use 4 bytes internal buffer instead of 8 (and fix a place where more that 4 bytes is currently needed). Zlib has 4 bytes checksum in footer and after that there will be no overshoot. Let me think a little bit more about that...

Regarding this great example.

`read` returns as much decompressed data as it has. We should not relay on that it reaches end of stream until it returns 0. Lets replace that with `readAll` and not test decompressor progress until it finishes. That leaves us with this problem:

```zig
test "problem" {
    const len: usize = 58;
    const compressed_with_four_extra_bytes = [len]u8{
        0x78, 0x9c, 0x73, 0xce, 0x2f, 0xa8, 0x2c, 0xca, 0x4c, 0xcf, 0x28, 0x51, 0x08, 0xcf, 0xcc, 0xc9,
        0x49, 0xcd, 0x55, 0x28, 0x4b, 0xcc, 0x53, 0x08, 0x4e, 0xce, 0x48, 0xcc, 0xcc, 0xd6, 0x51, 0x08,
        0xce, 0xcc, 0x4b, 0x4f, 0x2c, 0xc8, 0x2f, 0x4a, 0x55, 0x30, 0xb4, 0xb4, 0x34, 0xd5, 0xb5, 0x34,
        0x03, 0x00, 0x8b, 0x61, 0x0f, 0xa4, 0x52, 0x5a, 0x94, 0x12,
    };

    var base_stream = std.io.fixedBufferStream(compressed_with_four_extra_bytes[0..]);

    try std.testing.expectEqual(0, base_stream.getPos());
    var decompress_stream = std.compress.zlib.decompressor(base_stream.reader());
    var print_buf: [1024]u8 = undefined;

    try std.testing.expectEqual(8, base_stream.getPos());
    const got = try decompress_stream.reader().readAll(print_buf[0..]);

    try std.testing.expectEqual(46, got);
    try std.testing.expectEqualStrings("Copyright Willem van Schaik, Singapore 1995-96", print_buf[0..got]);

    // Expecting decompressor to not overshoot
    try std.testing.expectEqual(len - 4, base_stream.getPos()); // What I really expect
}
```

We have 4 bytes of overshoot after compressed chunk is consumed. And that is real problem for anyone who needs to consume from the same stream!

Here is a workarround by using stream which allows us to put back some bytes.

```zig
test "workaround" {
    const len: usize = 58;
    const compressed_with_four_extra_bytes = [len]u8{
        0x78, 0x9c, 0x73, 0xce, 0x2f, 0xa8, 0x2c, 0xca, 0x4c, 0xcf, 0x28, 0x51, 0x08, 0xcf, 0xcc, 0xc9,
        0x49, 0xcd, 0x55, 0x28, 0x4b, 0xcc, 0x53, 0x08, 0x4e, 0xce, 0x48, 0xcc, 0xcc, 0xd6, 0x51, 0x08,
        0xce, 0xcc, 0x4b, 0x4f, 0x2c, 0xc8, 0x2f, 0x4a, 0x55, 0x30, 0xb4, 0xb4, 0x34, 0xd5, 0xb5, 0x34,
        0x03, 0x00, 0x8b, 0x61, 0x0f, 0xa4, 0x52, 0x5a, 0x94, 0x12,
    };

    var base_stream = std.io.fixedBufferStream(compressed_with_four_extra_bytes[0..]);
    // Allows us to put back at most 8 bytes
    var lookahead_stream = std.io.bufferedTee(1024, 8, base_stream.reader(), std.io.null_writer);

    try std.testing.expectEqual(0, base_stream.getPos());
    var decompress_stream = std.compress.zlib.decompressor(lookahead_stream.reader());
    var print_buf: [1024]u8 = undefined;

    var got = try decompress_stream.reader().readAll(print_buf[0..]);
    try std.testing.expectEqual(46, got);
    try std.testing.expectEqualStrings("Copyright Willem van Schaik, Singapore 1995-96", print_buf[0..got]);

    // Rewind for unreadBytes; return 4 bytes overshoot to be read again
    lookahead_stream.putBack(decompress_stream.unreadBytes());
    // Read last 4 bytes
    got = try lookahead_stream.read(print_buf[0..]);
    try std.testing.expectEqual(4, got);
    try std.testing.expectEqualSlices(u8, compressed_with_four_extra_bytes[54..], print_buf[0..got]);
}
````

rockorager commented:
Checking in here, this went a bit over my head :).

Should we wait for a stdlib update for zlib or go ahead with the workaround?

mlarouche commented:
I think we should wait for the stdlib fix now that there's a PR for it :)

igor84 commented:
I am ok with implementing the "workaround" since it actually simplifies our code.

RossComputerGuy commented:
The PR just got merged into Zig, should be able to try it now with Zig master.

mlarouche commented:
I'm gonna try tomorrow with the newest zig master build, the latest available right now does not seems to have the fix

mlarouche commented:
I'm gonna merge as-is and submit the fix in another branch that make all the tests passes!
